tensorflow/contrib/training/python/training/sgdr_learning_rate_decay.py

```diff
@@ -0,0 +1,146 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+"""SGDR learning rate decay function."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import math
+
+from tensorflow.python.framework import constant_op
+from tensorflow.python.framework import ops
+from tensorflow.python.ops import math_ops, control_flow_ops
+
+
+def sgdr_decay(learning_rate, global_step, t_0=1000, t_mul=1, m_mul=1,
+               name=None):
+  """ This procedure implements Stochastic Gradient Descent with Warm
+  Restarts (SGDR) as described in "SGDR: Stochastic Gradient Descent
+  with Warm Restarts" by Ilya Loshchilov & Frank Hutter, Proceedings of
+  ICLR'2017, available at https://arxiv.org/pdf/1608.03983.pdf
+  The basic idea of the algorithm is the following.
+  The learning rate decreases according to cosine annealing:
+
+  ```python
+  learning_rate * 0.5 * (1 + cos(x_val * pi)) # for x_val defined in [0, 1]
+  ```
+
+  Thus, in the initial run (when the restart index i = 0),
+  the learning rate decreases from the initial learning rate
+  `learning_rate` (when `x_val=0`, we get `cos(0)=1`) to
+  0 (when `x_val=1`, we get `cos(pi)=-1`).
+  The decrease within i-th restart takes `t_i` steps,
+  while `t_0` is user-defined.
+  Then, we perform the first restart (i=1) by setting the learning rate to
+  `learning_rate*(m_mul^i)`, where `m_mul in [0,1]` (set to 1 by default).
+  Also, every restart runs for `t_i=t_0*(t_mul^i)` steps, i.e., every new
+  restart runs for `t_mul` longer than the previous one.
+
+  Importantly, when one has no access to a validation set, SGDR suggests
+  to report the best expected / recommended solution in the following way.
+  When we are within our initial run (i=0), every new solution represents
+  SGDR's recommended solution. Instead, when i>0, the recommended solution is
+  the one obtained at the end of each restart,
+  i.e., when the learning rate is 0.
+
+  Note that the minimum learning rate is set to 0 for simplicity,
+  you can adjust the code to deal with any positive minimum learning rate
+  as defined in the paper.
+
+  ```
+  Args:
+    learning_rate: A scalar `float32` or `float64` `Tensor` or a
+      Python number.  The initial learning rate.
+    global_step: A scalar `int32` or `int64` `Tensor` or a Python number.
+      Global step to use for the decay computation.  Must not be negative.
+    t_0: A scalar `int32` or `int64` `Tensor` or a Python number.
+      Must be positive.  Number of iterations in the first restart,
```

Comment:
We observe that all learning rate decay algorithms implemented in TensorFlow compute proper learning rate with respect to the global step.
This global step starts at 0 and is increased by one after each optimizer step (If we provide it as an argument when constructing optimizer).
So it would mean that t_0 is the duration of the first cycle measured as the number of minibatch updates (Iteration = minibatch update).
If one wants to use epochs, one should compute the number of iterations required for an epoch.
Example:
Minibatch size: 100
Training dataset size: 10 000
User wants the first decay cycle to span across 5 epochs: t_0 = 5 * 10000/100

We did it to be consistent with the other learning rate decay implementations available in TensorFlow. Original approach presented in the paper
however defines t_0 as the number of epochs. In both cases here and in the original implementation it is important to notice that the learning rate is adjusted after each minibatch update, so there is no sudden drop in the learning rate at the end of epoch.

We might need to clarify it inside the description.